### PR TITLE
Add SM4 block cipher [reboot]

### PR DESCRIPTION
**Pull request by levitte**
SM4 (formerly SMS4)[1] is a block cipher used in the Chinese
National Standard for Wireless LAN WAPI (Wired Authentication
and Privacy Infrastructure).
--from wikipedia

### Checklist

* [x] documentation is added or updated
* [x] tests are added or updated

-----

This is a reboot of #457, which seems to have been abandoned by its author

### CODE REVIEW

**Comment by levitte** [comment]
I believe this is ready for review

**Comment by levitte** [comment]
All fixups squashed too

**Comment by levitte** [comment]
Hmmmm, coveralls says that only 1 line in the new code is covered...  huh???

**Comment by sjaeckel** [comment]
> Hmmmm, coveralls says that only 1 line in the new code is covered... huh???

oopsie :D race condition on the cipher ID with the addition of DES-X via #587 :grimacing:

File: src/ciphers/sm4.c [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+#ifndef LTC_TEST
+    return CRYPT_NOP;
+#else
+    return sm4_self_test_ltc();
+#endif
+}
+
+const struct ltc_cipher_descriptor sm4_desc = {
+    "sm4",
+    27,
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
    28,
```

Comment (levitte) [reply src/ciphers/sm4.c]: Thank you.  Fixed

File: src/ciphers/sm4.c [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+        b[1] = (x >> 16) & 0xff;
+        b[2] = (x >> 8) & 0xff;
+        b[3] = (x >> 0) & 0xff;
+    } else {
+        b[0] = (x >> 0) & 0xff;
+        b[1] = (x >> 8) & 0xff;
+        b[2] = (x >> 16) & 0xff;
+        b[3] = (x >> 24) & 0xff;
+    }
+}
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]: Do we really need a runtime decision on that? We've the `LOAD` and `STORE` macros for that purpose.

File: src/ciphers/sm4.c [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+    a = (sm4_u8_t *)&A;
+    b = (sm4_u8_t *)&B;
+
+    b[0] = sbox(a[0]);
+    b[1] = sbox(a[1]);
+    b[2] = sbox(a[2]);
+    b[3] = sbox(a[3]);
+    return B;
+#endif
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]: Either we should remove one of the two options, or make it a real compile time decision (but only if there exists a specific reason why to use one or the other).

File: src/ciphers/sm4.c [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+    0xa0a7aeb5, 0xbcc3cad1, 0xd8dfe6ed, 0xf4fb0209,
+    0x10171e25, 0x2c333a41, 0x484f565d, 0x646b7279,
+};
+
+/*
+ * defined in section 7.3 (1) L'
+ */
+static sm4_u32_t _L(sm4_u32_t B)
+{
+    return B ^ rol(B, 13) ^ rol(B, 23);
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
    return B ^ ROLc(B, 13) ^ ROLc(B, 23);
```

File: src/ciphers/sm4.c [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+}
+
+/*
+   int sm4_setup(const unsigned char *key, int keylen, int num_rounds, symmetric_key *skey);
+   int sm4_ecb_encrypt(const unsigned char *pt, unsigned char *ct, const symmetric_key *skey);
+   int sm4_ecb_decrypt(const unsigned char *ct, unsigned char *pt, const symmetric_key *skey);
+   int sm4_test(void);
+   void sm4_done(symmetric_key *skey);
+   int sm4_keysize(int *keysize);
+   */
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
```

File: src/ciphers/sm4.c [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+    &sm4_keysize,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL,
+};
+
+#endif      /*LTC_SM4*/
+
+/* ref:         $Format:%D$ */
+/* git commit:  $Format:%H$ */
+/* commit time: $Format:%ai$ */
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
```

File: src/ciphers/sm4.c [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+    return B;
+#endif
+}
+
+/*
+ * defined in section 6.2 (2) Linear transformation L
+ */
+static sm4_u32_t L(sm4_u32_t B)
+{
+    return B ^ rol(B, 2) ^ rol(B, 10) ^ rol(B, 18) ^ rol(B, 24);
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
    return B ^ ROLc(B, 2) ^ ROLc(B, 10) ^ ROLc(B, 18) ^ ROLc(B, 24);
```

File: src/ciphers/sm4.c [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+    } while (0)
+
+#define result(x)                                       \
+    do {                                                \
+        sm4_printf("%s\n", (x) == 0 ? "PASS" : "FAIL"); \
+        fflush(stdout);                                 \
+    } while(0)
+
+
+void sm4_setkey(struct sm4_key *sm4, const unsigned char *key)
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
static void sm4_setkey(struct sm4_key *sm4, const unsigned char *key)
```

File: src/ciphers/sm4.c [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+#ifdef SM4DBG
+    sm4_printf("\n");
+#endif
+    for (i = 0; i < 32; ++i) {
+        X[i+4] = F(X+i, rk[i]);
+#ifdef SM4DBG
+        sm4_printf("rk[%02d] = %08X    X[%02d] = %08x\n",
+                   i, (unsigned int)(rk[i]), i+4, (unsigned int)(X[i+4]));
+#endif
+    }
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]:
```suggestion
    for (i = 0; i < 32; ++i) {
        X[i+4] = F(X+i, rk[i]);
    }
```

File: src/ciphers/sm4.c [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+    result = 0;
+
+    sm4_printf("SM4 Self Test\n");
+
+    sm4_setup(key, sizeof(key), 32, &skey);
+
+    /*A.1 example 1*/
+    test("single encryption");
+    sm4_ecb_encrypt(plaintext, output, &skey);
+    ret = sm4_memcmp(output, ciphertext, 16);
```
Comment (sjaeckel) [inline src/ciphers/sm4.c]: We should use `compare_testvector()` instead. Like that all the other invocations of `sm4_printf()`, `test()`, `result()` etc. can also go.

**Review by sjaeckel — CHANGES_REQUESTED** [review]

File: src/ciphers/sm4.c (hunk 3 above)
Comment (levitte) [reply src/ciphers/sm4.c]: I'm choosing the former, and use LOAD32H / STORE32H, to align with how it's done in other ciphers

File: src/ciphers/sm4.c (hunk 2 above)
Comment (levitte) [reply src/ciphers/sm4.c]: Removing them all

**Comment by sjaeckel** [comment]
What do you think of those ~two~ three commits? Nonsense or useful? Please rebase&squash the useful ones on top of develop, then I'll merge.

**Review by sjaeckel — APPROVED** [review]

**Comment by levitte** [comment]
They are all fine, so I adopted (squashed) them all

**Comment by levitte** [comment]
Now also rebased on a fresher develop

**Pull request merged**